### ISSUE TITLE
ZStream.tapSink hangs when the given sink doesn't consume the entire stream

### ISSUE THREAD

**Issue by eyalfa** [zio/zio#8081]
this program:
```
    val strm = ZStream
      .range(0, 100, 10)

    val sink0 = ZSink
      .head[Int]
      .mapZIO(opt => zio.Console.printLine("sink0: " + opt))


    strm
      .tapSink(sink0)
      .runForeach(i => zio.Console.printLine(s"run: $i"))
```

outputs the following and hangs:

```
run: 0
sink0: Some(0)
run: 1
run: 2
run: 3
run: 4
run: 5
run: 6
run: 7
run: 8
run: 9
run: 10
run: 11
run: 12
run: 13
run: 14
run: 15
run: 16
run: 17
run: 18
run: 19
```

I suspect the root cause for this issue is the bounded queue used in `ZStrea#tapSink`, it just lacks a mechanism for signaling downstream completion or failure, which causes upstream to keep attempting pushing into a full queue with no one draining it on the other side.

I've tried to fix this by using `ZSink#ignoreLeftover` but this only takes care of leftovers elements in the last consume chunk.

I was able to modify the sink to drain ALL leftovers:
```
  def streamHardDrainer[R, E, In, L, Z](sink : ZSink[R, E, In, L, Z]) : ZSink[R, E, In, L, Z] = {
    val ch: ZChannel[R, ZNothing, Chunk[In], Any, E, Chunk[L], Z] = sink
      .toChannel  <*
      ZChannel.identity[E, Chunk[In], Any].drain
    ch.toSink
  }
```

to summarise:
1. `ZStream#tapSink` should be able to track sink's completion and then stop enqueuing.
2. ZSink API can be enhanced to support 'hard' drain

**Comment by adamgfraser** [comment]
Thanks for reporting! Will take a look!

**Comment by eyalfa** [comment]
another approach for fixing this is to race the `offer` to the queue with `promise.await`, something like:

```
          def enqueue(t : zio.stream.Take[E, In]): ZIO[Any, Nothing, Unit] = {
            prom.await.race(q.offer(t).unit)
          }
```

will u accept a PR implementing this approach?
